src/org/traccar/protocol/OwnTracksProtocolDecoder.java

```diff
@@ -80,16 +87,34 @@ protected Object decode(
         if (root.containsKey("t")) {
             position.set("t", root.getString("t"));
         }
+        if (root.containsKey("p")) {
+            position.set("pb", root.getInt("p"));
+        }
         if (root.containsKey("batt")) {
             position.set(Position.KEY_BATTERY, root.getInt("batt"));
         }
-        if (root.containsKey("topic")) {
-            position.set("topic", root.getString("topic"));
+
+        if (root.containsKey("tst")) {
+            timestamp = root.getJsonNumber("tst").longValue();
+            if (timestamp < Integer.MAX_VALUE) {
+                timestamp *= 1000;
+            }
+        } else {
+            timestamp = new Date().getTime();
         }
+        position.setTime(new Date(timestamp));
 
-        position.setTime(new Date(root.getJsonNumber("tst").longValue() * 1000));
+        if (root.containsKey("tid")) {
+            deviceId = root.getString("tid");
+        }
+        if (root.containsKey("topic")) {
+            deviceId = root.getString("topic");
+            if (root.containsKey("tid")) {
```

Comment:
Not really. If `topic` exists, we want `tid` as an attribute, in addition to uniqueId set to `topic`.